mdp/utils/covariance.py

```diff
@@ -133,6 +135,120 @@ def fix(self, center=True):
         return cov_mtx, avg, tlen
 
 
+class UnevenlySampledCovarianceMatrix(CovarianceMatrix):
+    """This class stores an empirical covariance matrix that can be updated
+    incrementally. A call to the 'fix' method returns the current state of
+    the covariance matrix, the average and the number of observations, and
+    resets the internal data.
+
+    As compared to the *CovarianceMatrix* class, this class accepts unevenly
+    sampled input in conjunction with a time increment between samples. The
+    covariance matrix is then computed as a (centered) scalar product
+    between functions, that is sampled unevenly, using the trapezoid rule.
+
+    The mean is computed using the trapezoid rule as well.
+
+    Note that the internal sum is a standard __add__ operation.
+    """
+
+    def __init__(self, dtype=None):
+        """Initialize an *UnevenlySampledCovarianceMatrix*.
+
+        :param dtype: Datatype of the input.
+            Default is None. If dtype is not defined, it will be inherited
+            from the first data bunch received by 'update'. 
+        :type dtype: numpy.dtype or str
+        """
+        super(UnevenlySampledCovarianceMatrix,
+              self).__init__(dtype=dtype, bias=True)
+        # we need to keep steps and actual time
+        # values seperately, due to the weighing
+        # which is not required in the evenly sampled case
+        self._steps = 0
```

Comment:
I checked
`a=np.array([1.1, .2, 2.1])
 x= 0
 x+=np.sum(a)
`
in Python 2.7 and 3.5. In both it x returns as `3.4000000000000004`. But I agree setting it to be a floating point value is the more consistent and robust solution. I adapted it accordingly.

Reply:
Interesting. So far I thought `+=` couldn't change the type of a variable. Proved me wrong...

Reply:
Indeed:
```
stefan@stefan-y260 ~ $ python
Python 2.7.12 (default, Nov 12 2018, 14:36:49) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> x = 0
>>> type(x)
<type 'int'>
>>> id(x)
13939056
>>> x += .5
>>> x
0.5
>>> type(x)
<type 'float'>
>>> id(x)
14010040
>>> y = .0
>>> type(y)
<type 'float'>
>>> id(y)
14010016
>>> y += .5
>>> y
0.5
>>> type(y)
<type 'float'>
>>> id(y)
14009992
>>> 
```
Okay, ints and floats are not mutable and are always replaced by new instances on write access. That means, for a float or int, `+=` is actually not an in-place operation. If this operation happens very frequently it may be favorable to use a 1x1 numpy array instead so we get an actual inplace addition, saving lots of memory allocations.

Reply:
`import numpy as np
import time

n = 1000000

t = 0.
dt = 0.1
start = time.time()
for i in range(n):
    t += dt
print(time.time()-start)

t = np.array(0.)
dt = np.array(0.1)
start = time.time()
for i in range(n):
    t += dt
print(time.time()-start)`
`0.12778997421264648
2.427353858947754`

Reply:
Looks like Python's optimization is able to compete with numpy arrays for ordinary loops and a single element. Things should look different when using np.sum instead of the loop.
Anyway, it also occurred to me that the += on the int happens only once per chunk which is pretty seldom, so it's really the best to stick to the original implementation.
If you like, you can restore the original implementation without explicit `_tlen = 0.`.

Reply:
Is there a meaningful advantage? Would not intentions be more clear this way or do we rather save a line?

Reply:
This is just following the maxim to minimize redundant code. Since int and float are better compatible than I suspected, the reinitialization is not required. I remember some issues of this kind regarding numpy arrays, that's why I was suspicious here. Apparently that was a distinct situation.